### ISSUE TITLE
Presubmits Failing with "BRAM implementation not found for ROM memory."

### ISSUE THREAD

**Issue by gkelly** [lowRISC/opentitan#1430]
This started failing some time yesterday at the "Build bitstream with Vivado" step.

```
2020-01-27T18:57:09.9937896Z WARNING: [Vivado 12-180] No cells matched 'get_cells -hierarchical -filter { NAME =~  "*rom_rom*dout_o_reg_0" && PRIMITIVE_TYPE =~ BMEM.*.* }'.
2020-01-27T18:57:09.9938113Z ERROR: [Designcheck 2-3] BRAM implementation not found for ROM memory.
2020-01-27T18:57:09.9938330Z ERROR: [runtcl-1] ERROR: [Common 17-39] 'send_msg_id' failed due to earlier errors.
2020-01-27T18:57:09.9938539Z ERROR: [Common 17-39] 'send_msg_id' failed due to earlier errors.
```

**Comment by towoe** [comment]
Is this not the strange error message if the rom init file can not be found?

top_earlgrey_nexysvideo.core parameter ROM_INIT_FILE

**Comment by imphil** [comment]
I'm looking at it, not sure yet where the problem came from now (the changes between good and bad commits are not really pointing at anything obvious).

**Comment by imphil** [comment]
OK, I roughly found what's going on. Working on a fix now.

**Comment by imphil** [comment]
Status:
- We end up with a slightly messed-up passing of the defines to Vivado (a missing space), which causes the bootrom VMEM file not to be found by Vivado.
- I cannot yet reproduce it locally, or see any commits in fusesoc/edalize (or other changes) that would cause that.
- I'm unable to continue the investigation today, so either merge if this single task fails for today, or (better) avoid merging today and I'll fix this first thing in the (UK) morning tomorrow.

**Comment by towoe** [comment]
I did a diff on the `runme.log` of my local build and the azure run.
Looks like the parameters are not set correctly?

Here is an example for the first difference:
```diff
@@ -1,10 +1,5 @@
-INFO: [Synth 8-6157] synthesizing module 'prim_clock_gating' [$LOCAL/build-out/hw/src/lowrisc_prim_clock_gating_0/abstract/prim_clock_gating.sv:12]
-	Parameter Impl bound to: 1 - type: integer 
-INFO: [Synth 8-6157] synthesizing module 'prim_xilinx_clock_gating' [$LOCAL/build-out/hw/src/lowrisc_prim_xilinx_clock_gating_0/rtl/prim_xilinx_clock_gating.sv:5]
-INFO: [Synth 8-6157] synthesizing module 'BUFGCE' [/opt/xilinx/Vivado/2018.3/scripts/rt/data/unisim_comp.v:619]
-	Parameter CE_TYPE bound to: SYNC - type: string 
-	Parameter IS_CE_INVERTED bound to: 1'b0 
-	Parameter IS_I_INVERTED bound to: 1'b0 
-INFO: [Synth 8-6155] done synthesizing module 'BUFGCE' (1#1) [/opt/xilinx/Vivado/2018.3/scripts/rt/data/unisim_comp.v:619]
-INFO: [Synth 8-6155] done synthesizing module 'prim_xilinx_clock_gating' (2#1) [$LOCAL/build-out/hw/src/lowrisc_prim_xilinx_clock_gating_0/rtl/prim_xilinx_clock_gating.sv:5]
-INFO: [Synth 8-6155] done synthesizing module 'prim_clock_gating' (3#1) [$LOCAL/build-out/hw/src/lowrisc_prim_clock_gating_0/abstract/prim_clock_gating.sv:12]
+INFO: [Synth 8-6157] synthesizing module 'prim_clock_gating' [$AZURE/hw/src/lowrisc_prim_clock_gating_0/abstract/prim_clock_gating.sv:12]
+	Parameter Impl bound to: 0 - type: integer 
+INFO: [Synth 8-6157] synthesizing module 'prim_generic_clock_gating' [/home/agent/runner/_work/1/a/build-out/hw/src/lowrisc_prim_generic_clock_gating_0/rtl/prim_generic_clock_gating.sv:7]
+INFO: [Synth 8-6155] done synthesizing module 'prim_generic_clock_gating' (1#1) [/home/agent/runner/_work/1/a/build-out/hw/src/lowrisc_prim_generic_clock_gating_0/rtl/prim_generic_clock_gating.sv:7]
+INFO: [Synth 8-6155] done synthesizing module 'prim_clock_gating' (2#1) [/home/agent/runner/_work/1/a/build-out/hw/src/lowrisc_prim_clock_gating_0/abstract/prim_clock_gating.sv:12]
```

And then the "generic" primitives are used. This is the same for the rom, which I guess does not offer setting the values by the parameter.

**Comment by towoe** [comment]
Locally I use fusesoc 1.9.4 as 1.9.3 does not work, most likely a local problem with arch/python3.8 and me not knowing how to work with `[sudo] pip3 [--user] [-e]`.
I will run the steps in a container tomorrow morning @imphil

**Comment by towoe** [comment]
My guess is that something is wrong with FuseSoC.

The file `synth-vivado/lowrisc_systems_top_earlgrey_nexysvideo_0.1.tcl`
contains in azure
`set_property verilog_define {PRIM_DEFAULT_IMPL=prim_pkg::ImplGenericROM_INIT_FILE=$BIN_DIR/sw/device/fpga/boot_rom/boot_rom.vmem} [get_filesets sources_1]`

in my local system
`set_property verilog_define {PRIM_DEFAULT_IMPL=prim_pkg::ImplXilinx ROM_INIT_FILE=$BIN_DIR/sw/device/fpga/boot_rom/boot_rom.vmem } [get_filesets sources_1]`

and in some container
`set_property verilog_define {PRIM_DEFAULT_IMPL=prim_pkg::ImplXilinxROM_INIT_FILE=../../../../../build-bin/sw/device/fpga/boot_rom/boot_rom.vmem} [get_filesets sources_1]`

Notice the difference in the space between the defines and the different content for `PRIM_DEFAULT_IMPL`.

**Comment by imphil** [comment]
Thanks to more debugging by @towoe we now know what's going on. Here's the short story:

fusesoc relies on edalize to write EDA tool-specific TCL files, e.g. to create Vivado project files. These files are templated using a templating engine called Jinja2. Two days ago, Jinja2 released a new version 2.11.0 which changed the whitespace stripping behavior slightly. This resulted in a TCL file to create a Vivado project that looked like that:

```
set_property verilog_define {PRIM_DEFAULT_IMPL=prim_pkg::ImplXilinxROM_INIT_FILE=/home/agent/runner/_work/1/a/build-bin/sw/device/fpga/boot_rom/boot_rom.vmem} [get_filesets sources_1]
```

Note the missing spaces between the defines. This meant that defines (and parameters) weren't properly passed to Vivado, which then in turn couldn't find the ROM_INIT_FILE, and synthesis failed.


* Edalize PR with workaround: https://github.com/olofk/edalize/pull/98
* Upstream bug report: https://github.com/pallets/jinja/issues/1138


Things we've learned:
- We should pin Python dependencies to a specific version to avoid a mismatch between versions used in CI, and versions used on developer's machines. I've opened https://github.com/lowRISC/opentitan/issues/1444 for that.
- Our guard scripts that fail synthesis if no ROM file is found work properly.
